content/6_Advanced/Max_Flow.mdx

```diff
@@ -112,6 +112,140 @@ int main() {
 
 <Problems problems="match" />
 
+It's recommended that you solve the first problem - Download Speed - in the section before trying this one.
+
+### Solution 1
+
+A bipartite graph doesn't seem like to have anything in common with a flow network, but we can shift our point of view just by
+adding the source connected to the nodes in set $U$ and the sink connected to the nodes in set $V$. And that's all.
+Now we have a flow network where every capacity is equal to 1.
+
+We transformed our bipartite graph into a network flow so the maximum network flow is equal
+to the maximum matching.
+Now, we can apply our favorite max flow algorithm to solve the problem!
+
+**Time Complexity: $\mathcal{O}(VE^2)$**
+
+### Solution 2
+
+However, we can do better than this.
+First let's define some properties of matching algorithms.
+
+Let's say the set $M$ contains all the edges that the maximum matching consists of.
+
+We define an **alternating path** as a path whose edges are in the matching, $M$, and not in the matching, in an alternating fashion. An alternating path stars with an unmatched node and ends once it cannot append another edge while maintaining an alternating sequence.
+
+An **augmenting path** is built upon the alternating path and unmatched nodes at both ends - the nodes are not included in $M$.
+
+Tha maximum matching $M$ can be further improved if and only if an augmenting path is found in $M$, otherwise $M$ is the maximum matching. It may seem difficult to understand, but the main idea is: ``The maximum matching can be further improved as long as the alternating sequence can be extended ``. For a better understanding, you can imagine the shoelaces as an alternating path in a bipartite graph - or the sneakers.
+
+The algorithm described above is called the [Hopcroft-Karp algorithm](https://en.wikipedia.org/wiki/Hopcroft%E2%80%93Karp_algorithm).
+
+Here's an animation of the algorithm if you're still a bit confused:
+
+<video width="960" height="720" controls>
+	<source src="/animations/Hopcroft-Karp.mp4" type="video/mp4"/>
+</video>
+
+**Time Complexity: $\mathcal{O}(E\sqrt{V})$**
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int INF = 1e9;
+
+int n, m, k;
+// dist[node] = the position of node in the alternating path
+vector<int> Pair, dist;
```

Comment:
can you rename `Pair` to something else so it isn't an eyesore

Reply:
bump

Reply:
Alright done